dali/kernels/tensor_view.h

```diff
@@ -0,0 +1,237 @@
+  TensorView<Backend, DataType, other_dim> to_static() {
+    static_assert(other_dim != DynamicTensorShape,
+                  "Conversion to static only allowed for static shape");
+    return {data, shape.to_static<other_dim>()};
+  }
+
+  template <int other_dim>
+  TensorView<Backend, DataType, other_dim> to_static(const TensorShape<other_dim> &new_shape) {
+    static_assert(other_dim != DynamicTensorShape,
+                  "Conversion to static only allowed for static shape");
+    return {data, new_shape};
+  }
+
+  template <int other_dim>
+  TensorView<Backend, DataType, other_dim> to_static(TensorShape<other_dim> &&new_shape) {
+    static_assert(other_dim != DynamicTensorShape,
+                  "Conversion to static only allowed for static shape");
+    return {data, std::move(new_shape)};
+  }
+
+  DataType *data;
+  TensorShape<DynamicTensorShape> shape;
+
+  static const bool has_static_dim = false;
+  int dim() const { return shape.size(); }
+
+  // static const int static_dim = dim_;
+  // int dim() const { return ShapeDim(shape); }
+
+  // template <typename Offset>
+  // DataType *at(const Offset &pos) const {
+  //   return data + CalcOfffset(shape, pos);
+  // }
+
+  // template <typename... Indices>
+  // DataType *at(int64_t idx0, Indices &&... idx) const {
+  //   return data + CalcOfffset(shape, {idx0, (static_cast<int64_t>(idx))...});
+  // }
+};
+
+template <typename Backend, typename DataType, int dim_>
+struct TensorView {
+  TensorView() : data(nullptr), shape() {}
+  TensorView(DataType *data, TensorShape<dim_> shape) : data(data), shape(shape) {}
+  TensorView(const TensorView &) = default;
+  TensorView(TensorView &&) = default;
+  TensorView &operator=(const TensorView &) = default;
+  TensorView &operator=(TensorView &&) = default;
+
+  DataType *data;
+  TensorShape<dim_> shape;
+
+  static const bool has_static_dim = false;
+  constexpr int dim() const { return shape.size(); }
+
+  // static const int static_dim = dim_;
+  // int dim() const { return ShapeDim(shape); }
+
+  // template <typename Offset>
+  // DataType *at(const Offset &pos) const {
+  //   return data + CalcOfffset(shape, pos);
+  // }
+
+  // template <typename... Indices>
+  // DataType *at(int64_t idx0, Indices &&... idx) const {
+  //   return data + CalcOfffset(shape, {idx0, (static_cast<int64_t>(idx))...});
+  // }
+};
+
+template <typename Backend, typename DataType, int sample_dim_>
+struct TensorListView;
+
+template <typename Backend, typename DataType, int sample_dim_>
+struct TensorListView {
+  TensorListView() : data(nullptr), shape(), offsets() {}
+  TensorListView(DataType *data, const std::vector<TensorShape<sample_dim_>> &shapes) : data(data), shape(shapes), offsets(calculate_offsets(shape)) {}
+
+  DataType *data;
+  TensorListShape<sample_dim_> shape;
+  std::vector<ptrdiff_t> offsets;
+
+  template <int dim = sample_dim_>
+  TensorView<Backend, DataType, dim> operator[](int64_t sample) const {
```

Comment:
This template here is unusable. How would you provide the explicit template argument? Even if it's possible, it would be extremely awkward. Please provide an alternate method (`at`?) for this kind of access.

Reply:
It now allows to access only sample_ndim TensorViews.